### ISSUE TITLE
Transform2 not calling TranformUpdated

### ISSUE THREAD

**Issue by rafaelalmeidatk** [craftworkgames/MonoGame.Extended#507]
Transform2 has [two events](https://github.com/craftworkgames/MonoGame.Extended/blob/develop/Source/MonoGame.Extended/Transform.cs#L106) but only the `TransformBecameDirty` is being called. I placed breakpoints at `RecalculateLocalMatrixIfNecessary` and `RecalculateWorldMatrixIfNecessary`, they are never reachable, I am changing the Position property of the transform.

**Comment by LithiumToast** [comment]
You said the event `TransformBecameDirty` is getting invoked. This means that the method `WorldMatrixBecameDirty` must be getting called, like you said, from the setter of the `Position` property. The matrices only get recalculated if necessary, i.e., the method `GetLocalMatrix` or `GetWorldMatrix` is called and the flags are dirty. If you don't call these methods, nothing will happen, as intended.

**Comment by rafaelalmeidatk** [comment]
Yes, right, `WorldMatrixBecameDirty` it is being called and `GetLocalMatrix`/`GetWorldMatrix` clears the dirty state, but acessing the `Position` property should not also call `RecalculateLocalMatrixIfNecessary`/`RecalculateWorldMatrixIfNecessary` to recalculate the position if need?

If I have this:

    if (transform.IsPositionDirty())
    {
        Console.WriteLine("dirty");
        var pos = transform.WorldPosition;
    }

The output is only printed once, since `WorldPosition` returns a property from `WorldMatrix` and in the getter of `WorldMatrix` it calls `RecalculateWorldMatrixIfNecessary`. But if I have this:

    if (transform.IsPositionDirty())
    {
        Console.WriteLine("dirty");
        var pos = transform.Position;
    }

The output is printed every frame. Is this the desired behaviour?

**Comment by LithiumToast** [comment]
No. `Transform2.Position` is the *local* position, like it says in the XML docs. If you want the *world* position use `Transform2.WorldPosition`.

**Comment by rafaelalmeidatk** [comment]
That makes sense. So do I use `Position` to move the transform and `WorldPosition` to get the world position to render it and clear the dirty flag?

**Comment by LithiumToast** [comment]
Yes you should use `WorldPosition` for drawing. This means the matrix is only recalculated if needed before it is drawn. For drawing sprites, unfortunately, `SpriteBatch` wants the decomposed matrix. It would make more sense to write a drop-in replacement for `SpriteBatch` that you could pass in the `WorldMatrix` directly instead of position, rotation, etc. `Matrix2` is already pretty optimized (it's only 6 floats), so it's not much different then passing 2 floats for position, 1 float for rotation, 2 floats for scale, and 2 floats for origin (7 floats total). However, `SpriteBatch` would then have to do a transform for the 4 vertices of the quad from local (model) space to world space. I can't really comment on the performance penalty (or relative gain?) without some empirical data. It would be a simpler API though with less arguments to pass for drawing...

**Comment by craftworkgames** [comment]
> When did Matrix2D change to Matrix2?

Now that we're in moving towards 2.0 I took the opportunity to rename some classes to make the library more consistent.

>  and why is it called Matrix2?

Honestly, I renamed a whole bunch of things at once and I didn't put a lot of thought into this particular rename. So, my mistake on that one. It's not too late to change it again though, so let's do that once we've figured out if we're renaming this class or getting rid of it in favor of the one's from .NET Core.

> If you are using .NET Core, which I assume you are, then why not use that instead to reduce the liability of MonoGame.Extended? 

Very good point. I didn't know about those classes in .NET Core. Moving everything to .NET Standard / Core only happened very recently and I'm sure there's still plenty of room for improvements. It's only thanks to the help from guys like yourself that we can make these changes. 

> Take a look at the source code for Matrix3x2 and compare it to the source code for Matrix2.

The source code does indeed look quite similar. I'll have to take a closer look before removing Matrix2 but considering we only added Matrix2D because there wasn't anything like that in the .NET framework BCL then I think it makes sense.

**Comment by LithiumToast** [comment]
I wrote the original `Matrix2D` with help from Nez's `Matrix2D` lol

There are some methods which I added to `Matrix2D` which are not there in `Matrix3x2` but those can easily be extension methods.

**Comment by rafaelalmeidatk** [comment]
>It would make more sense to write a drop-in replacement for SpriteBatch

Unfortunately, this is out of my skills yet, but I can search more about it. The `SpriteBatch` is good enough for the begin of the game, though.

So should I use the solution 3? It won't require a change at the lib, I am just worried about the lambda that is never cleared from the event.

**Comment by LithiumToast** [comment]
```
transform.TransformBecameDirty += () =>
     {
         Console.WriteLine("dirty");
         collider.HandleTransformUpdate(transform);
     };
```

This anon method will create a closure over the `collider` and `transform`. This essentially means a new object will be created at runtime to hold these variables for the anon method.
